### ISSUE TITLE
Hold pending check for 10 seconds / provide means of explicitly overriding the transactional install cleanup

### ISSUE THREAD

**Issue by pauxus** [chocolatey/choco#822]
**Original title:** no longer able to run choco commands from chocolateyInstall.ps1

Since 0.9.10, with the cleanup of unfinished packages, it is no longer possible to run choco commands from inside a chocolateyInstall.ps1

Our rationale for that is that we have a kickstarter package, which resides on an uprotected feed. This package asks for username and password and configures the chocolatey sources to point to our feed aggregator (nexus), using the user credentials.

This stopped working with 0.9.10, since any run of the choco command seems to clean the currently installing package from the lib folder, resulting in an unsuccessful installation.

In order to fix this, we would either need to:

- provide a way to tell commands not to clean up pending packages
- mark a package a "uncleanable"
- exempt the cleanup from certain unrelated commands (like choco source)
- provide cmdlets for common tasks to use form inside an install script (i.e. changeChocoSource or something)

**Comment by ferventcoder** [comment]
This is an interesting side effect of ensuring a pending install.

The way to handle this is to delete the `.pending` file in the package directory.

**Comment by ferventcoder** [comment]
> * provide a way to tell commands not to clean up pending packages

Right now the answer is to delete the `.pending` file in the top of the package directory prior to calling any choco commands.

> * mark a package a "uncleanable"

Interesting

> * exempt the cleanup from certain unrelated commands (like choco source)

This could be confusing in what commands would trigger the pending cleanup

> * provide cmdlets for common tasks to use form inside an install script (i.e. changeChocoSource or something)

This could be the best option. 


This is quite an interesting way of going about it. Another option is to look at the timestamp for that .pending file and only react if it is over say 2 minutes old?

**Comment by pauxus** [comment]
The workaround works for now. However, just for anyone needing this in the meantime:

The filename is `.chocolateyPending`, not `.pending`

**Comment by ferventcoder** [comment]
Yes, apologies.

**Comment by ferventcoder** [comment]
Little early here ha 👍

**Comment by ferventcoder** [comment]
After thinking about this some more, 2 minutes is way too long. 20 seconds is more reasonable be default. I think we will also allow you to drop a file to skip pending. Does that seem reasonable?

**Comment by ferventcoder** [comment]
FYI, one of the things we are moving towards is only allowing one choco process at a time.

**Comment by ferventcoder** [comment]
Include an empty `.chocolateyPendingSkip` in the package.

**Comment by ferventcoder** [comment]
This will be released in 0.9.10.3

**Comment by masaeedu** [comment]
From the command line, is it possible to start multiple chocolatey install scripts without interfering with each other? The current default of nuking the in-progress install seems unreasonable. Either:

- the new invocation of choco install should proceed without trying to delete the in-progress package. If either fails due to MSI reservation conflicts so be it
- or the new invocation should wait until the pending installation is complete before doing anything, and should notify the user it is waiting
- or an error should be produced informing the user a pending install is in progress and no action will be taken
- or the user should be prompted to make an explicit decision to abort and remove the pending install

"Pending" installs and their removal isn't as clean as it may seem at first glance, since choco has no idea what artifacts the package has already inserted into the system, and unceremoniously removing the package does not even allow the user the ability to run the package's uninstall script to try and clean up. In most cases it just ends up irrevocably borking the package on that system.

**Comment by ferventcoder** [comment]
@masaeedu can you give me a really good use case for running choco in parallel?

**Comment by ferventcoder** [comment]
What you are stating here would probably be best as a new "Improve pending operations" enhancement ticket.

**Comment by masaeedu** [comment]
A product is being installed via chocolatey. I open a terminal and try to check what packages are installed:

```
choco list -lo
```

This dumps the existing install, but not before it has proceeded far enough that reattempts to install the package fail. I cannot uninstall because the package is not installed. I cannot force install because that doesn't work. I must know what the package does to the system and undo it manually, or dump and rebuild the system.

Substitute listing packages with trying to install process explorer, checking the version of a package, enabling a chocolatey feature etc. etc. etc.

**Comment by ferventcoder** [comment]
* `choco list -lo` - this could be incorrect because an install is happening right then.

I can see other things that maybe should not trigger pending checks, but other installs would. Again, Please create a new ticket with the title "Improve Package Pending Operations" - perhaps one thing we can do is put a process lock on the file and if a lock is found, we simply skip that pending operation.

**Comment by ferventcoder** [comment]
@pauxus I think you are going to love what we did in #943.
